AdvancedSystems.Connector/Services/MsSqlServerConnectionService.cs

```diff
@@ -19,45 +26,213 @@ public MsSqlServerConnectionService(ILogger<MsSqlServerConnectionService> logger
+
+    public string ConnectionString { get; private set; }
+
+    public ConnectionState ConnectionState { get; private set; }
+
+    #endregion
+
+    #region Helpers
+
+    private void ConnectionStateHandler(object sender, StateChangeEventArgs e)
     {
-        var builder = new SqlConnectionStringBuilder
-        {
-            UserID = this._settings.UserID,
-            Password = this._settings.Password,
-            DataSource = this._settings.DataSource,
-            InitialCatalog = this._settings.InitialCatalog,
-            MinPoolSize = this._settings.MinPoolSize,
-            MaxPoolSize = this._settings.MaxPoolSize,
-            ConnectTimeout = this._settings.ConnectTimeout,
-            CommandTimeout = this._settings.CommandTimeout,
-            TrustServerCertificate = this._settings.TrustServerCertificate,
-            ApplicationName = this._settings.ApplicationName,
-            Encrypt = this._settings.Encrypt,
-        };
-
-        return builder.ConnectionString;
+        this.ConnectionState = e.CurrentState;
     }
 
-    #region Properties
+    private void InfoMessageHandler(object sender, SqlInfoMessageEventArgs e)
+    {
+        if (e.Errors.Count == 0) return;
 
-    public string ConnectionString { get; private set; }
+        byte warningThreshold = 10;
+        byte errorThreshold = 20;
+        SqlError lastError = e.Errors[^1];
+
+        if (lastError.Class <= warningThreshold)
+        {
+            this._logger.LogWarning("{Server} issued the following warning: {Message}.", lastError.Server, lastError.Message);
+        }
+        else if (lastError.Class > warningThreshold && lastError.Class <= errorThreshold)
+        {
+            this._logger.LogError("{Server} raised an error on line {Line}: {Message} (State={State}).", lastError.Server, lastError.LineNumber, lastError.Message, lastError.State);
+        }
+        else
+        {
+            // A severity over 20 causes the connection to close
+            string reason = "Connection was closed";
+            this._logger.LogCritical("{Reason} ({Message}).", reason, lastError.Message);
+            throw new DbConnectionException($"{reason} ({lastError.Message}).");
+        }
+    }
 
     #endregion
 
     #region Methods
 
-    public void ExecuteQuery()
+    public DataSet ExecuteQuery(IDatabaseCommand databaseCommand)
     {
-        throw new NotImplementedException();
+        DataSet result = new();
+
+        try
+        {
+            using var connection = new SqlConnection(this.ConnectionString);
+            connection.StateChange += ConnectionStateHandler;
+            connection.InfoMessage += InfoMessageHandler;
+            connection.Open();
+
+            using var sqlCommand = connection.CreateCommand();
+            sqlCommand.CommandText = databaseCommand.CommandText;
+            sqlCommand.CommandType = databaseCommand.CommandType.Cast();
+
+            if (!databaseCommand.Parameters.IsNullOrEmpty())
```

Comment:
Use extension method from `AdvancedSystems.Core`

Reply:
No longer necessary for now, since this method got removed with the last refactoring commit